memcached.c

```diff
@@ -535,17 +538,54 @@ void conn_worker_readd(conn *c) {
     if (c->state == conn_closing) {
         drive_machine(c);
         return;
+    } else if (c->state == conn_io_queue) {
+        // machine will know how to return based on secondary state.
+        drive_machine(c);
+    } else {
+        conn_set_state(c, conn_new_cmd);
     }
-    c->state = conn_new_cmd;
+}
 
-#ifdef EXTSTORE
-    // If we had IO objects, process
-    if (c->io_wraplist) {
-        //assert(c->io_wrapleft == 0); // assert no more to process
-        conn_set_state(c, conn_mwrite);
-        drive_machine(c);
+void conn_io_queue_add(conn *c, int type, void *ctx, io_queue_stack_cb cb, io_queue_stack_cb com_cb, io_queue_cb fin_cb) {
+    io_queue_t *q = c->io_queues;
+    while (q->type != IO_QUEUE_NONE) {
+        q++;
+    }
+    q->type = type;
+    q->ctx = ctx;
+    q->stack_ctx = NULL;
+    q->submit_cb = cb;
+    q->complete_cb = com_cb;
+    q->finalize_cb = fin_cb;
+    return;
+}
+
+io_queue_t *conn_io_queue_get(conn *c, int type) {
+    io_queue_t *q = c->io_queues;
+    while (q->type != IO_QUEUE_NONE) {
+        if (q->type == type) {
+            return q;
+        }
+        q++;
+    }
+    return NULL;
+}
+
+// called after returning to the main worker thread.
+// users of the queue need to distinguish if the IO was actually consumed or
+// not and handle appropriately.
+static void conn_io_queue_complete(conn *c) {
+    io_queue_t *q = c->io_queues;
+    while (q->type != IO_QUEUE_NONE) {
+        // Reuse the same submit stack. We zero it out first so callbacks can
+        // queue new IO's if necessary.
+        if (q->stack_ctx) {
+            void *tmp = q->stack_ctx;
+            q->stack_ctx = NULL;
+            q->complete_cb(q->ctx, tmp);
+        }
+        q++;
     }
```

Comment:
should we call memset on the queue we processed after calling the complete_cb.

Reply:
I think I see it now..
since the same queue structure is used for all extstore specific code for the life of the connection, this doesn't need to be zeroed out and it will be reused. That makes sense. That's why stack_ctx is set to null to make it usable for the next IO to be submitted.

Reply:
Yes the queues are "global" (per worker thread) structures.

Reply:
Thank you. Did you mean to say per worker thread or per connection structures?

Reply:
I guess the callbacks are global but each connection gets a queue. All the funcs are labelled `conn_io_etc` where appropriate. ie; independent connections will all have the same queue types with the same callback funcs, but individual stack pointers for each connection.

Reply:
Thanks, that makes sense.